doc/reference/trd104-syscalls.md

```diff
@@ -509,19 +509,36 @@ kernel returns `Failure` with an error code of `NODEVICE`.
 ---------------------------------
 
 The Read-Write Allow system call class is how a userspace process
-shares a buffer with the kernel that the kernel can read and write. 
+shares a buffer with the kernel that the kernel can read and write.
 
 Calling a Read-Write Allow system call returns a buffer (address and
-length).  On the first call to a Read-Write Allow system call, the
-kernel returns a zero-length buffer. Subsequent successful calls to
-Read-Write Allow return the previous buffer passed.  The standard
-access model for allowed buffers is that userspace does not read or
-write a buffer that has been allowed: access to the memory is intended
-to be exclusive either to userspace or to the kernel. To regain access
-to a passed buffer, the process calls the same Read-Write Allow system
-call again, which, if successful, returns the buffer. It can do so
-with a zero-length buffer if it wishes to pass no memory to the
-kernel. 
+length). Kernel drivers MUST be implemented such that the returned
+buffer follows a defined set of rules, as specified in the following:
+
+- On the first successful call to a Read-Write Allow system call, the
```

Comment:
This seems a bit redundant with lines 554-561 below, but also not completely in sync. E.g., 561 says the address MUST be 0.

Perhaps it would be better to distinguish system call implementations from the kernel. E.g., system call implementations MUST do this but the kernel doesn't check it, so userspace libraries SHOULD? This text says the kernel MUST follow the rules but that it doesn't check them, which seems a bit weird.